src/Microsoft.ML.Core/Utilities/DoubleParser.cs

```diff
@@ -535,13 +533,19 @@ private static bool TryParseSpecial(ReadOnlySpan<char> span, ref int ich, out Si
             return false;
         }
 
-        private static bool TryParseCore(ReadOnlySpan<char> span, ref int ich, ref bool neg, ref ulong num, ref long exp)
+        private static bool TryParseCore(ReadOnlySpan<char> span, ref int ich, ref bool neg, ref ulong num, ref long exp, OptionFlags flags = OptionFlags.Default)
         {
             Contracts.Assert(0 <= ich & ich <= span.Length);
             Contracts.Assert(!neg);
             Contracts.Assert(num == 0);
             Contracts.Assert(exp == 0);
 
+            char decimalMarker;
+            if ((flags & OptionFlags.UseCommaAsDecimalMarker) != 0)
+                decimalMarker = ',';
+            else
+                decimalMarker = '.';
```

Comment:
Why as a flag `OptionFlags.UseCommaAsDecimalMarker` instead of just having a user enter a char directly? Having user enter a char directly lets them any character like `_` in `1000_23`.

Pandas lets users enter the char directly, see `decimal` parameter:
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html

They also allow for entering the char for `thousands` (technically allow a string).

Excel also lets the user enter the chars directly: 
https://www.officetooltips.com/excel_2016/tips/change_the_decimal_point_to_a_comma_or_vice_versa.html

Reply:
The public API to let users change the decimal marker was implemented and decided on #5145. We accept a character as the DecimalMarker option, but we actually only accept "," or "." and throw exceptions if the user sets other character.

On this PR I am not changing that behavior, but actually only changing how we connect the TextLoader to the DoubleParser (as the way we connected them wasn't thread safe, and would have unexpected behavior in some cases). Since we only accept "," or ".", then I used a Flag here. Since this flag is of internal use (i.e. it isn't determining a public API), it could later be replaced by an options object or something else if we get to accept other characters/strings as decimal separators. In any case, if we were to accept other characters different from "," and ".", then we'd actually need to change other things on the parser and on textloader to make this work correctly... so perhaps that could be left for future work, since I don't think we'd have time to correctly implement and test those other options before the next release. So, I think it's better to simply merge this PR to enable users to use "," and "." as decimal markers in a safe manner in the upcoming release, and then discuss about the possibility of enabling other characters.

As for the thousands parameter you mention, it is included inside a list of possible features that we might want to implement for TextLoader, but it isn't in our plans to implement it before the upcoming release.

---
In reply to: [429606655](https://github.com/dotnet/machinelearning/pull/5154#discussion_r429606655) [](ancestors = 429606655)